Store userid in metadata

iPraveenParihar opened the pull request:
<!-- Please take a look at our [Contributing](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#Code-contribution-workflow)
documentation before submitting a Pull Request!
Thank you for contributing to ceph-csi! -->

# Describe what this PR does #

## Problem

Without a mechanism to track nodeID:userID mappings per volume, it becomes challenging to:

Monitor which userIDs are actively being used across the cluster
Make informed decisions about key rotation, allowing safe cleanup of old keys

## Proposed Solution
The solution involves tracking userID information by storing it in the metadata of RBD images and CephFS subvolumes using the key: .[rbd|cephfs].csi.ceph.com/userID/<NodeId>:<userID>

refer design doc ([here](https://github.com/ceph/ceph-csi/pull/5425/))for more details: 
- https://github.com/ceph/ceph-csi/pull/5425/

**Checklist:**

* [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#commit-messages).
* [x] Reviewed the developer guide on [Submitting a Pull Request](https://github.com/ceph/ceph csi/blob/devel/docs/development-guide.md#development-workflow)
* [x] [Pending release notes](https://github.com/ceph/ceph-csi/blob/devel/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next major release.
* [ ] Documentation has been updated, if necessary.
* [ ] Unit tests have been added, if necessary.
* [ ] Integration tests have been added, if necessary.

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

* `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)

</details>

nixpanic commented on internal/cephfs/driver.go:
```diff
@@ -127,8 +127,10 @@ func (fs *Driver) Run(conf *util.Config) {
-	fs.cd = csicommon.NewCSIDriver(conf.DriverName, util.DriverVersion, conf.NodeID, conf.InstanceID,
-		conf.EnableFencing)
+	fs.cd = csicommon.NewCSIDriver(conf.DriverName, util.DriverVersion, conf.NodeID, conf.InstanceID)
+	fs.cd.SetEnableFencing(conf.EnableFencing)
```
Instead of having `Set...` functions for everything, consider passing options like `WithFencing` to `NewCSIDriver()`. Similar to how gRPC CallOptions work, that makes it a little more dynamic and easier extendible.

iPraveenParihar replied:
addressed now, Thanks for the suggestion.

nixpanic reviewed (commented):
surely there should be e2e tests for this?

Madhu-1 commented on cmd/cephcsi.go:
```diff
@@ -70,6 +70,7 @@ func init() {
 	flag.StringVar(&conf.ClusterName, "clustername", "", "name of the cluster")
 	flag.BoolVar(&conf.SetMetadata, "setmetadata", true, "set metadata on the volume")
 	flag.BoolVar(&conf.EnableFencing, "enable-fencing", false, "enable fencing of nodes during non-graceful shutdowns")
+	flag.BoolVar(&conf.SetUserIDMapping, "setuseridmapping", false, "set userID node mapping on volume metadata")
```
i dont think we need to introduce a flag for this one. we already have `setmetadata` flag for it.

iPraveenParihar replied:
we are removing the metadata in ControllerUnpublishVolume() and existing PVs will not be having controller secrets and may be breaking change?

iPraveenParihar replied:
We need to advertise the ControllerPublish/UnpublishVolume, so some knob is needed based on which we need to advertise?

Madhu-1 replied:
Breaking change means it breaks the existing functionality, We can document it as its a minor release to adopt to it. I would suggest to enable it by default and document it in release notes. This is just a metadata stored in the PVC.  This is true for other metadata as well it can have same state when PV is reattached to other PVC, unless they are running the omap generator (which not everyone runs)

Rakshith-R replied:
> Breaking change means it breaks the existing functionality, We can document it as its a minor release to adopt to it. I would suggest to enable it by default and document it in release notes. This is just a metadata stored in the PVC. This is true for other metadata as well it can have same state when PV is reattached to other PVC, unless they are running the omap generator (which not everyone runs)

@Madhu-1 If we don't remove it in controllerUnpublish, it will be stale data on the image.

We should either have both the nodestage and controllerUnpublish parts working or completely disabled.

Madhu-1 replied:
As i explain there are two things, 1. we will document it in the release notes and update the storageclass as well. and  there are change for other stale data as well. IMO it should be fine. if they think they are having stale data or doesnt want the feature, they can disable the metadata flag(which we already have one), We should try to avoid introducing new flags for new features unless it causes some major problems.

Rakshith-R replied:
IMO, nodeID:userID mapping is meant to imply that the volume is actively mapped/mounted on to node <nodeID> with ceph user <userID> 
https://github.com/ceph/ceph-csi/blob/devel/docs/design/proposals/userID-mapping.md

Without a setUserID flag and fencing flag disabled, CephCSI will end up only adding <nodeID:userID> and never removing it.

@Madhu-1 WDYT about basing the decision of setting userID too on the `EnableFencing` flag ?

or have a common flag for both setUserID and enable fencing ?

@nixpanic can you please add your opinions as well.

Madhu-1 replied:
Both are different functionality and it doesnt depend on each other, We can check if enablemetadata flag is set and publish the caps from the driver. I  would still tie this to enable-metadata flag not with the fencing Flag, This is just a metadata stored on the image or subvolume. It will not have any impact on the functionality. Again as i explained above there are chance for stale data for PVC and namespace name in the metadata as well.

Rakshith-R replied:
>We can check if enablemetadata flag is set and publish the caps from the driver.

This works for me
@iPraveenParihar please make these changes.

iPraveenParihar replied:
@Madhu-1, Today we are passing the `setmetadata` flag only to controller plugin and NOT to nodeplugin.
Will need this flag to be set for nodeplugin as well.

iPraveenParihar replied:
Or flag will be used only to publish caps and not for set/remove nodeId:userId mapping metadata?

Rakshith-R replied:
Yes. Pass the flag to nodeplugn but by default it is true so we should be good.

Madhu-1 commented on PendingReleaseNotes.md:
```diff
@@ -8,6 +8,8 @@
 - rbd: add support for [CSI Snapshot Metadata Service RPCs](https://github.com/container-storage-interface/spec/blob/master/spec.md#snapshot-metadata-service-rpcs)
 - feature: handle non graceful node shutdown [PR](https://github.com/ceph/ceph-csi/pull/5429/)
    - refer design doc for more details - [here](docs/design/proposals/non-graceful-node-shutdown.md)
+- feature: set nodeId:userId mapping metadata [PR](https://github.com/ceph/ceph-csi/pull/5445)
```
This is already feature section. we dont need to add `feature:` here

Madhu-1 commented on internal/cephfs/controllerserver.go:
```diff
@@ -1178,14 +1178,79 @@ func (cs *ControllerServer) ControllerUnpublishVolume(
+// removeUserIdMapping attempts to remove nodeId:userId mapping metadata from the subvolume.
+//
+// Parameters:
+//   - nodeId: The ID of the node that may be fenced.
```
fenced?

iPraveenParihar replied:
typo, addressed.

Madhu-1 commented on internal/cephfs/controllerserver.go:
```diff
@@ -1178,14 +1178,79 @@ func (cs *ControllerServer) ControllerUnpublishVolume(
+	if !cs.Driver.IsUserIdMappingEnabled() {
+		return nil
+	}
+	if secrets == nil {
+		var vi util.CSIIdentifier
+		err := vi.DecomposeCSIID(volumeId)
+		if err != nil {
+			return fmt.Errorf("error decoding volume ID (%s): %w", volumeId, err)
+		}
+
+		secrets, err = util.GetControllerPublishSecret(vi.ClusterID, util.CephFsType)
+		if err != nil {
+			log.WarningLog(ctx, "controller publish secret not found: %v", err)
+
+			return nil
+		}
+	}
+	credentials, err := util.NewAdminCredentials(secrets)
+	if err != nil {
+		return fmt.Errorf("failed to create credentials from secrets: %w", err)
+	}
+	defer credentials.DeleteCredentials()
+
+	volOptions, _, err := store.NewVolumeOptionsFromVolID(ctx, volumeId, nil, secrets, cs.ClusterName, cs.SetMetadata)
+	if err != nil {
+		return fmt.Errorf("failed to generate volume from volume ID %s: %w", volumeId, err)
+	}
+	defer volOptions.Destroy()
+
+	volClient := core.NewSubVolume(volOptions.GetConnection(), &volOptions.SubVolume, volOptions.ClusterID,
+		cs.ClusterName, cs.SetMetadata)
```
This is kind of getting duplication in multiple function Please use helpfer as required and pass the volClient dont create it always in internal functions

iPraveenParihar replied:
refactored

Madhu-1 commented on internal/csi-common/driver.go:
```diff
@@ -101,6 +103,18 @@ func (d *CSIDriver) IsFencingEnabled() bool {
 	return d.enableFencing
 }
 
+func (d *CSIDriver) SetEnableFencing(value bool) {
```
value is too generate please use better names

iPraveenParihar replied:
removed now,

Madhu-1 commented on internal/csi-common/driver.go:
```diff
@@ -101,6 +103,18 @@ func (d *CSIDriver) IsFencingEnabled() bool {
+	d.enableFencing = value
+}
+
+func (d *CSIDriver) SetUserIdMapping(value bool) {
+	d.setUserIdMapping = value
+}
+
+func (d *CSIDriver) IsUserIdMappingEnabled() bool {
```
add comments for exported functions

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -1736,14 +1736,79 @@ func (cs *ControllerServer) ControllerUnpublishVolume(
+		return errors.New("nodeId cannot be empty")
+	}
+
+	if secrets == nil {
+		var vi util.CSIIdentifier
+		err := vi.DecomposeCSIID(volumeId)
+		if err != nil {
+			return fmt.Errorf("failed to decode volume ID (%s): %w", volumeId, err)
+		}
+
+		secrets, err = util.GetControllerPublishSecret(vi.ClusterID, util.RBDType)
+		if err != nil {
+			log.WarningLog(ctx, "controller publish secret not found: %v", err)
+
+			return nil
+		}
+	}
+	credentials, err := util.NewAdminCredentials(secrets)
+	if err != nil {
+		return fmt.Errorf("failed to create credentials from secrets: %w", err)
+	}
+	defer credentials.DeleteCredentials()
+
+	rv, err := GenVolFromVolID(ctx, volumeId, credentials, secrets)
+	if err != nil {
+		return fmt.Errorf("failed to generate volume from volume ID %q: %w",
+			volumeId, err)
+	}
```
Please avoid code duplications

iPraveenParihar replied:
refactored.

iPraveenParihar commented:
Thanks @Madhu-1, for the early reviews.
Am working on avoiding duplications.

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.31

iPraveenParihar commented:
@Rakshith-R @nixpanic PTAL

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.31

Madhu-1 commented on e2e/pod.go:
```diff
@@ -722,3 +731,86 @@ func verifyReadAffinity(
+			metadataKey, metadataValue, expectedValue)
+	}
+
+	err = deletePVCAndApp("", f, pvc, app)
```
delete pod and verify that the key is removed

Madhu-1 commented on e2e/rbd.go:
```diff
@@ -529,6 +531,14 @@ var _ = Describe("RBD", func() {
+				if err != nil {
+					framework.Failf("failed to verify userId mapping metadata exists: %v", err)
+				}
+			})
```
validate that no stale resources

Madhu-1 replied:
this is still missing stale resource validation

Madhu-1 commented on e2e/cephfs.go:
```diff
@@ -360,6 +362,24 @@ var _ = Describe(cephfsType, func() {
+				err = deleteResource(cephFSExamplePath + "storageclass.yaml")
+				if err != nil {
+					framework.Failf("failed to delete CephFS storageclass: %v", err)
+				}
```
validate no stale resouces

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -1702,12 +1702,32 @@ func (cs *ControllerServer) ControllerPublishVolume(
+	}
+	defer credentials.DeleteCredentials()
+
+	rv, err := GenVolFromVolID(ctx, volumeId, credentials, secrets)
```
rv.Destory is missing?

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -1731,26 +1751,82 @@ func (cs *ControllerServer) ControllerUnpublishVolume(
+	}
+	defer credentials.DeleteCredentials()
+
+	rv, err := GenVolFromVolID(ctx, volumeId, credentials, secrets)
```
rv.Destory is missing

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.31

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.31

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.31

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.31

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.31

Madhu-1 commented on PendingReleaseNotes.md:
```diff
@@ -6,8 +6,10 @@
-- feature: handle non graceful node shutdown [PR](https://github.com/ceph/ceph-csi/pull/5429/)
+- handle non graceful node shutdown [PR](https://github.com/ceph/ceph-csi/pull/5429/)
    - refer design doc for more details - [here](docs/design/proposals/non-graceful-node-shutdown.md)
+- set nodeId:userId mapping metadata [PR](https://github.com/ceph/ceph-csi/pull/5445)
```
nodeId:userId mapping  in metadata

Madhu-1 commented on e2e/pod.go:
```diff
@@ -722,3 +731,124 @@ func verifyReadAffinity(
+
+	// verify userId mapping metadata is removed after pod deletion
+	// retry up to 3 times with 10-second intervals
+	maxRetries := 3
+	retryInterval := 10 * time.Second
```
@iPraveenParihar why do we need retry here? i assume the pod will only gets deleted after successful ControllerUnpublish. Do you mean after we delete the metadata its not removed in ceph for sometime?

iPraveenParihar replied:
Yea, I observed a race where by the time we check for the metadata, ControllerUnpublishVolume() call is not been called and metadata is not yet removed. So thought of adding retry here

Madhu-1 replied:
This is not as per expectation, Thats seems to be a problem, Did you trace the logs to see what is happening? Do we need to adjust the E2E ?

iPraveenParihar replied:
VolumeAttachment is not immediately deleted right after pod is deleted. Probably, there is some time gap.

I see small time difference between pod deleted and ControllerUnpublishVolume call

we check for metadata exists @ 11:39:21 (06:09:23)
```
I0728 11:39:21.507860 2846093 pod.go:425] Waiting for pod csi-rbd-demo-pod to be deleted
I0728 11:39:23.514792 2846093 exec_util.go:63] ExecWithOptions {Command:[/bin/sh -c rbd image-meta get replicapool/csi-vol-a37658af-1884-40d5-97c0-fb0747eda9d1 .rbd.csi.ceph.com/userID/dr1] Namespace:rook-ceph PodName:rook-ceph-tools-7b75b967db-xxvj5 ContainerName:rook-ceph-tools Stdin:<nil> CaptureStdout:true CaptureStderr:true PreserveWhitespace:true Quiet:false}
```

ControllerUnpublishVolume call @ 06:09:28 (roughly after 5 seconds)
```
I0728 06:09:28.390784       1 utils.go:341] ID: 37 GRPC call: /csi.v1.Controller/ControllerUnpublishVolume
I0728 06:09:28.391436       1 utils.go:342] ID: 37 GRPC request: {"node_id":"dr1","secrets":"***stripped***","volume_id":"0001-0024-2ab5c1ca-0cf3-46c3-a42e-b9fdfc33bae0-0000000000000003-a37658af-1884-40d5-97c0-fb0747eda9d1"}
I0728 06:09:28.395367       1 omap.go:89] ID: 37 got omap values: (pool="replicapool", namespace="", name="csi.volume.a37658af-1884-40d5-97c0-fb0747eda9d1"): map[csi.imageid:3b853c6822b81 csi.imagename:csi-vol-a37658af-1884-40d5-97c0-fb0747eda9d1 csi.volname:pvc-fa3e6359-e089-421b-a23e-f9fc10de648f csi.volume.owner:rbd-719]
I0728 06:09:28.494642       1 controllerserver.go:1821] ID: 37 successfully removed user ID mapping for nodeId dr1
I0728 06:09:28.494846       1 utils.go:348] ID: 37 GRPC response: {}
```

Madhu-1 replied:
Can you please open a tracker for this one, we might need to update our tests for pod deletion?

iPraveenParihar replied:
Yea, we need to adjust our E2E, other tests are also failing.

For example,  test `verify generic ephemeral volume suppor`

```
  [38;5;9m[FAILED] backend images not matching kubernetes resource count,image count 1 kubernetes resource count 0
  backend image Info:
   [csi-vol-0e182ca2-a61b-4f44-9e68-9c6c7967c5f6]
   images information and status Pool: replicapool, Image: csi-vol-0e182ca2-a61b-4f44-9e68-9c6c7967c5f6, Info: {"name":"csi-vol-0e182ca2-a61b-4f44-9e68-9c6c7967c5f6","id":"137925ae9ace","size":1073741824,"objects":256,"order":22,"object_size":4194304,"snapshot_count":0,"block_name_prefix":"rbd_data.137925ae9ace","format":2,"features":["layering"],"op_features":[],"flags":[],"create_timestamp":"Mon Jul 28 06:16:48 2025","access_timestamp":"Mon Jul 28 06:16:48 2025","modify_timestamp":"Mon Jul 28 06:16:48 2025"}
  , Status: {"watchers":[]}
  ```

But I don't see the /ControllerUnpublishVolume and /DeleteVolume call from the logs

```
  I0728 06:16:48.666406       1 utils.go:341] ID: 47 Req-ID: pvc-ec9fe03c-7ada-41fd-a552-429bbd5e3a7f GRPC call: /csi.v1.Controller/CreateVolume
  I0728 06:16:48.666523       1 utils.go:342] ID: 47 Req-ID: pvc-ec9fe03c-7ada-41fd-a552-429bbd5e3a7f GRPC request: {"capacity_range":{"required_bytes":1073741824},"name":"pvc-ec9fe03c-7ada-41fd-a552-429bbd5e3a7f","parameters":{"clusterID":"bbbc7815-0990-49f8-9336-441f115a5cc9","csi.storage.k8s.io/pv/name":"pvc-ec9fe03c-7ada-41fd-a552-429bbd5e3a7f","csi.storage.k8s.io/pvc/name":"csi-rbd-demo-ephemeral-pod-mypvc","csi.storage.k8s.io/pvc/namespace":"rbd-8446","imageFeatures":"layering","pool":"replicapool"},"secrets":"***stripped***","volume_capabilities":[{"access_mode":{"mode":"SINGLE_NODE_WRITER"},"mount":{"fs_type":"ext4","mount_flags":["discard"]}}]}
  I0728 06:16:48.666747       1 rbd_util.go:1430] ID: 47 Req-ID: pvc-ec9fe03c-7ada-41fd-a552-429bbd5e3a7f setting disableInUseChecks: false image features: [layering] mounter: rbd
  I0728 06:16:48.667452       1 omap.go:89] ID: 47 Req-ID: pvc-ec9fe03c-7ada-41fd-a552-429bbd5e3a7f got omap values: (pool="replicapool", namespace="", name="csi.volumes.default"): map[]
  I0728 06:16:48.672640       1 omap.go:159] ID: 47 Req-ID: pvc-ec9fe03c-7ada-41fd-a552-429bbd5e3a7f set omap keys (pool="replicapool", namespace="", name="csi.volumes.default"): map[csi.volume.pvc-ec9fe03c-7ada-41fd-a552-429bbd5e3a7f:0e182ca2-a61b-4f44-9e68-9c6c7967c5f6])
  I0728 06:16:48.675430       1 omap.go:159] ID: 47 Req-ID: pvc-ec9fe03c-7ada-41fd-a552-429bbd5e3a7f set omap keys (pool="replicapool", namespace="", name="csi.volume.0e182ca2-a61b-4f44-9e68-9c6c7967c5f6"): map[csi.imagename:csi-vol-0e182ca2-a61b-4f44-9e68-9c6c7967c5f6 csi.volname:pvc-ec9fe03c-7ada-41fd-a552-429bbd5e3a7f csi.volume.owner:rbd-8446])
  I0728 06:16:48.675456       1 rbd_journal.go:522] ID: 47 Req-ID: pvc-ec9fe03c-7ada-41fd-a552-429bbd5e3a7f generated Volume ID (0001-0024-bbbc7815-0990-49f8-9336-441f115a5cc9-0000000000000002-0e182ca2-a61b-4f44-9e68-9c6c7967c5f6) and image name (csi-vol-0e182ca2-a61b-4f44-9e68-9c6c7967c5f6) for request name (pvc-ec9fe03c-7ada-41fd-a552-429bbd5e3a7f)
  I0728 06:16:48.675508       1 rbd_util.go:470] ID: 47 Req-ID: pvc-ec9fe03c-7ada-41fd-a552-429bbd5e3a7f rbd: create replicapool/csi-vol-0e182ca2-a61b-4f44-9e68-9c6c7967c5f6 size 1024M (features: [layering]) using mon rook-ceph-mon-a.rook-ceph.svc.cluster.local:6789
  I0728 06:16:48.675534       1 rbd_util.go:1694] ID: 47 Req-ID: pvc-ec9fe03c-7ada-41fd-a552-429bbd5e3a7f setting image options on replicapool/csi-vol-0e182ca2-a61b-4f44-9e68-9c6c7967c5f6
  I0728 06:16:48.689493       1 controllerserver.go:808] ID: 47 Req-ID: pvc-ec9fe03c-7ada-41fd-a552-429bbd5e3a7f created image replicapool/csi-vol-0e182ca2-a61b-4f44-9e68-9c6c7967c5f6 backed for request name pvc-ec9fe03c-7ada-41fd-a552-429bbd5e3a7f
  I0728 06:16:48.705791       1 omap.go:159] ID: 47 Req-ID: pvc-ec9fe03c-7ada-41fd-a552-429bbd5e3a7f set omap keys (pool="replicapool", namespace="", name="csi.volume.0e182ca2-a61b-4f44-9e68-9c6c7967c5f6"): map[csi.imageid:137925ae9ace])
  I0728 06:16:48.771517       1 utils.go:348] ID: 47 Req-ID: pvc-ec9fe03c-7ada-41fd-a552-429bbd5e3a7f GRPC response: {"volume":{"capacity_bytes":1073741824,"volume_context":{"clusterID":"bbbc7815-0990-49f8-9336-441f115a5cc9","imageFeatures":"layering","imageName":"csi-vol-0e182ca2-a61b-4f44-9e68-9c6c7967c5f6","journalPool":"replicapool","pool":"replicapool"},"volume_id":"0001-0024-bbbc7815-0990-49f8-9336-441f115a5cc9-0000000000000002-0e182ca2-a61b-4f44-9e68-9c6c7967c5f6"}}
  I0728 06:16:50.545114       1 utils.go:341] ID: 48 GRPC call: /csi.v1.Controller/ControllerPublishVolume
  I0728 06:16:50.545200       1 utils.go:342] ID: 48 GRPC request: {"node_id":"minikube","secrets":"***stripped***","volume_capability":{"access_mode":{"mode":"SINGLE_NODE_WRITER"},"mount":{"fs_type":"ext4","mount_flags":["discard"]}},"volume_context":{"clusterID":"bbbc7815-0990-49f8-9336-441f115a5cc9","imageFeatures":"layering","imageName":"csi-vol-0e182ca2-a61b-4f44-9e68-9c6c7967c5f6","journalPool":"replicapool","pool":"replicapool","storage.kubernetes.io/csiProvisionerIdentity":"1753683286984-5062-rbd.csi.ceph.com"},"volume_id":"0001-0024-bbbc7815-0990-49f8-9336-441f115a5cc9-0000000000000002-0e182ca2-a61b-4f44-9e68-9c6c7967c5f6"}
  I0728 06:16:50.546449       1 omap.go:89] ID: 48 got omap values: (pool="replicapool", namespace="", name="csi.volume.0e182ca2-a61b-4f44-9e68-9c6c7967c5f6"): map[csi.imageid:137925ae9ace csi.imagename:csi-vol-0e182ca2-a61b-4f44-9e68-9c6c7967c5f6 csi.volname:pvc-ec9fe03c-7ada-41fd-a552-429bbd5e3a7f csi.volume.owner:rbd-8446]
  I0728 06:16:50.565404       1 utils.go:348] ID: 48 GRPC response: {}
  ```

As said in earlier (https://github.com/ceph/ceph-csi/pull/5445#discussion_r2234856942) there is a time gap and at places where we check for backend images count will fail. We need to account for the delay.

I suggest to move the e2e from this PR and tackle the problem in a separate PR?

@Madhu-1 @Rakshith-R ^^

iPraveenParihar replied:
I tried few adjustments to e2e, but that messes up with other tests or will need changes at all places.

Madhu-1 replied:
@iPraveenParihar i dont think its a good idea to address the E2E in followup PR. As a thumb rule, E2E is to ensure everything works and no regression happens. If something in E2E(new/old) is failing because of this new change we need to identify and address in the same PR.

Madhu-1 commented on charts/ceph-csi-cephfs/values.yaml:
```diff
@@ -382,6 +382,8 @@ storageClass:
   provisionerSecretNamespace: ""
   controllerExpandSecret: csi-cephfs-secret
   controllerExpandSecretNamespace: ""
+  controllerPublishSecret: csi-cephfs-secret
+  controllerPublishSecretNamespace: ""
```
update the same in storageclass in example folder as well.

iPraveenParihar replied:
Added already.

Madhu-1 reviewed (commented):
small nits and few previous comments need to be addressed in E2E, LGTM

Rakshith-R commented:
```
  [1mSTEP:[0m verify generic ephemeral volume support [38;5;243m@ 07/25/25 05:51:50.576[0m
  I0725 05:51:50.583166   81715 warnings.go:110] "Warning: would violate PodSecurity \"restricted:latest\": allowPrivilegeEscalation != false (container \"web-server\" must set securityContext.allowPrivilegeEscalation=false), unrestricted capabilities (container \"web-server\" must set securityContext.capabilities.drop=[\"ALL\"]), runAsNonRoot != true (pod or container \"web-server\" must set securityContext.runAsNonRoot=true), seccompProfile (pod or container \"web-server\" must set securityContext.seccompProfile.type to \"RuntimeDefault\" or \"Localhost\")"
  I0725 05:51:50.583296 81715 pod.go:375] Waiting up to csi-rbd-demo-ephemeral-pod to be in Running state
  I0725 05:51:58.588403 81715 exec_util.go:63] ExecWithOptions {Command:[/bin/sh -c rbd ls --format=json --pool=replicapool] Namespace:rook-ceph PodName:rook-ceph-tools-7dd7bbcd4b-qjh79 ContainerName:rook-ceph-tools Stdin:<nil> CaptureStdout:true CaptureStderr:true PreserveWhitespace:true Quiet:false}
  I0725 05:51:58.588452 81715 exec_util.go:68] ExecWithOptions: Clientset creation
  I0725 05:51:58.588498 81715 exec_util.go:84] ExecWithOptions: execute(https://192.168.49.2:8443/api/v1/namespaces/rook-ceph/pods/rook-ceph-tools-7dd7bbcd4b-qjh79/exec?command=%2Fbin%2Fsh&command=-c&command=rbd+ls+--format%3Djson+--pool%3Dreplicapool&container=rook-ceph-tools&stderr=true&stdout=true)
  I0725 05:51:58.706162 81715 exec_util.go:201] unexpected error trying to use websockets for pod exec: <nil>
  I0725 05:51:58.708750 81715 exec_util.go:63] ExecWithOptions {Command:[/bin/sh -c rados ls --pool=replicapool | grep -v default | grep -v csi.volume.group. |  grep -c ^csi.volume.] Namespace:rook-ceph PodName:rook-ceph-tools-7dd7bbcd4b-qjh79 ContainerName:rook-ceph-tools Stdin:<nil> CaptureStdout:true CaptureStderr:true PreserveWhitespace:true Quiet:false}
  I0725 05:51:58.708791 81715 exec_util.go:68] ExecWithOptions: Clientset creation
  I0725 05:51:58.708840 81715 exec_util.go:84] ExecWithOptions: execute(https://192.168.49.2:8443/api/v1/namespaces/rook-ceph/pods/rook-ceph-tools-7dd7bbcd4b-qjh79/exec?command=%2Fbin%2Fsh&command=-c&command=rados+ls+--pool%3Dreplicapool+%7C+grep+-v+default+%7C+grep+-v+csi.volume.group.+%7C++grep+-c+%5Ecsi.volume.&container=rook-ceph-tools&stderr=true&stdout=true)
  I0725 05:51:58.839196 81715 exec_util.go:201] unexpected error trying to use websockets for pod exec: <nil>
  I0725 05:51:58.841730 81715 exec_util.go:63] ExecWithOptions {Command:[/bin/sh -c rados listomapkeys csi.volumes.default --pool=replicapool | wc -l] Namespace:rook-ceph PodName:rook-ceph-tools-7dd7bbcd4b-qjh79 ContainerName:rook-ceph-tools Stdin:<nil> CaptureStdout:true CaptureStderr:true PreserveWhitespace:true Quiet:false}
  I0725 05:51:58.841774 81715 exec_util.go:68] ExecWithOptions: Clientset creation
  I0725 05:51:58.841840 81715 exec_util.go:84] ExecWithOptions: execute(https://192.168.49.2:8443/api/v1/namespaces/rook-ceph/pods/rook-ceph-tools-7dd7bbcd4b-qjh79/exec?command=%2Fbin%2Fsh&command=-c&command=rados+listomapkeys+csi.volumes.default+--pool%3Dreplicapool+%7C+wc+-l&container=rook-ceph-tools&stderr=true&stdout=true)
  I0725 05:51:59.021463 81715 exec_util.go:201] unexpected error trying to use websockets for pod exec: <nil>
  I0725 05:51:59.027592 81715 pod.go:425] Waiting for pod csi-rbd-demo-ephemeral-pod to be deleted
  I0725 05:52:01.032780 81715 pvc.go:464] waiting for PVC mypvc in state &PersistentVolumeClaimStatus{Phase:,AccessModes:[],Capacity:ResourceList{},Conditions:[]PersistentVolumeClaimCondition{},AllocatedResources:ResourceList{},AllocatedResourceStatuses:map[ResourceName]ClaimResourceStatus{},CurrentVolumeAttributesClassName:nil,ModifyVolumeStatus:nil,} to be deleted (0 seconds elapsed)
  I0725 05:52:01.035296 81715 exec_util.go:63] ExecWithOptions {Command:[/bin/sh -c rbd ls --format=json --pool=replicapool] Namespace:rook-ceph PodName:rook-ceph-tools-7dd7bbcd4b-qjh79 ContainerName:rook-ceph-tools Stdin:<nil> CaptureStdout:true CaptureStderr:true PreserveWhitespace:true Quiet:false}
  I0725 05:52:01.035334 81715 exec_util.go:68] ExecWithOptions: Clientset creation
  I0725 05:52:01.035380 81715 exec_util.go:84] ExecWithOptions: execute(https://192.168.49.2:8443/api/v1/namespaces/rook-ceph/pods/rook-ceph-tools-7dd7bbcd4b-qjh79/exec?command=%2Fbin%2Fsh&command=-c&command=rbd+ls+--format%3Djson+--pool%3Dreplicapool&container=rook-ceph-tools&stderr=true&stdout=true)
  I0725 05:52:01.153385 81715 exec_util.go:201] unexpected error trying to use websockets for pod exec: <nil>
  I0725 05:52:01.156333 81715 exec_util.go:63] ExecWithOptions {Command:[/bin/sh -c rbd info --pool=replicapool csi-vol-45c73401-92dd-4b1c-829f-0c9c412f1c4b --format json] Namespace:rook-ceph PodName:rook-ceph-tools-7dd7bbcd4b-qjh79 ContainerName:rook-ceph-tools Stdin:<nil> CaptureStdout:true CaptureStderr:true PreserveWhitespace:true Quiet:false}
  I0725 05:52:01.156370 81715 exec_util.go:68] ExecWithOptions: Clientset creation
  I0725 05:52:01.156416 81715 exec_util.go:84] ExecWithOptions: execute(https://192.168.49.2:8443/api/v1/namespaces/rook-ceph/pods/rook-ceph-tools-7dd7bbcd4b-qjh79/exec?command=%2Fbin%2Fsh&command=-c&command=rbd+info+--pool%3Dreplicapool+csi-vol-45c73401-92dd-4b1c-829f-0c9c412f1c4b+--format+json&container=rook-ceph-tools&stderr=true&stdout=true)
  I0725 05:52:01.290525 81715 exec_util.go:201] unexpected error trying to use websockets for pod exec: <nil>
  I0725 05:52:01.293092 81715 exec_util.go:63] ExecWithOptions {Command:[/bin/sh -c rbd status --pool=replicapool csi-vol-45c73401-92dd-4b1c-829f-0c9c412f1c4b --format json] Namespace:rook-ceph PodName:rook-ceph-tools-7dd7bbcd4b-qjh79 ContainerName:rook-ceph-tools Stdin:<nil> CaptureStdout:true CaptureStderr:true PreserveWhitespace:true Quiet:false}
  I0725 05:52:01.293144 81715 exec_util.go:68] ExecWithOptions: Clientset creation
  I0725 05:52:01.293181 81715 exec_util.go:84] ExecWithOptions: execute(https://192.168.49.2:8443/api/v1/namespaces/rook-ceph/pods/rook-ceph-tools-7dd7bbcd4b-qjh79/exec?command=%2Fbin%2Fsh&command=-c&command=rbd+status+--pool%3Dreplicapool+csi-vol-45c73401-92dd-4b1c-829f-0c9c412f1c4b+--format+json&container=rook-ceph-tools&stderr=true&stdout=true)
  I0725 05:52:01.426293 81715 exec_util.go:201] unexpected error trying to use websockets for pod exec: <nil>
  [38;5;9m[FAILED][0m in [It] - /go/src/github.com/ceph/ceph-csi/e2e/rbd.go:248 [38;5;243m@ 07/25/25 05:52:01.426[0m
  I0725 05:52:01.452337 81715 log.go:56] Logs of cephcsi-e2e-1bb98910/csi-rbdplugin-provisioner-648cbbf9bf-cgh9q:csi-rbdplugin on node minikube
  ````
  e2e failure seems to be due some network flaky issue in another test case.
  unrelated to this pr
  
  https://jenkins-ceph-csi.apps.ocp.cloud.ci.centos.org/blue/rest/organizations/jenkins/pipelines/mini-e2e_k8s-1.31/runs/597/nodes/90/log/?start=0

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.32

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.32

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.32

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.32

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.32

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.32

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.32

iPraveenParihar commented:
failed [log](https://jenkins-ceph-csi.apps.ocp.cloud.ci.centos.org/blue/rest/organizations/jenkins/pipelines/mini-e2e_k8s-1.32/runs/288/nodes/90/log/?start=0)
```
  [38;5;9m[FAILED] failed to validate fuse mount options: failed to get app: client rate limiter Wait returned an error: rate: Wait(n=1) would exceed context deadline[0m
  [38;5;9mIn [1m[It][0m[38;5;9m at: [1m/go/src/github.com/ceph/ceph-csi/e2e/cephfs.go:430[0m [38;5;243m@ 07/28/25 16:36:43.047[0m
  ```

iPraveenParihar commented:
/retest ci/centos/mini-e2e/k8s-1.32

iPraveenParihar commented:
Failed again due to [logs](https://jenkins-ceph-csi.apps.ocp.cloud.ci.centos.org/blue/organizations/jenkins/mini-e2e_k8s-1.32/detail/mini-e2e_k8s-1.32/290/pipeline/)
```
  [38;5;9m[FAILED] failed to validate fuse mount options: failed to get app: client rate limiter Wait returned an error: rate: Wait(n=1) would exceed context deadline[0m
  [38;5;9mIn [1m[It][0m[38;5;9m at: [1m/go/src/github.com/ceph/ceph-csi/e2e/cephfs.go:430[0m [38;5;243m@ 07/28/25 18:52:34.147[0m
```

iPraveenParihar commented:
/retest ci/centos/mini-e2e/k8s-1.32

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.32

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.31

iPraveenParihar commented:
/retest ci/centos/mini-e2e-helm/k8s-1.32

iPraveenParihar commented:
failures are not related to this PR changes. 
Same failure observed as https://github.com/ceph/ceph-csi/pull/5454#issuecomment-3130923787

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.32

iPraveenParihar commented:
[logs](https://jenkins-ceph-csi.apps.ocp.cloud.ci.centos.org/blue/rest/organizations/jenkins/pipelines/mini-e2e-helm_k8s-1.32/runs/266/nodes/100/log/?start=0)

```
  [38;5;9m[FAILED] failed to set cluster name: timeout waiting for clustername arg update cephcsi-e2e-84a4d9826a5c/csi-cephfsplugin-provisioner: deployment cephcsi-e2e-84a4d9826a5c/csi-cephfsplugin-provisioner did not become available yet: client rate limiter Wait returned an error: rate: Wait(n=1) would exceed context deadline[0m
```

```
  [38;5;9m[FAILED] timeout waiting for deployment csi-rbdplugin-provisioner: error waiting for deployment "csi-rbdplugin-provisioner" status to match desired state: [0m
  [38;5;9mIn [1m[BeforeEach][0m[38;5;9m at: [1m/go/src/github.com/ceph/ceph-csi/e2e/rbd.go:408[0m [38;5;243m@ 07/29/25 08:54:14.585[0m
  ```

This has become frequent error 😢 , let me try with increasing the 
```
ClientQPS:   40,
ClientBurst: 100,
```
for framework

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.32

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.32

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.32

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.32

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.32

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.32

iPraveenParihar commented:
test - `create a storageclass with pool and a PVC then bind it to an app`


Never seen this [failure)(https://jenkins-ceph-csi.apps.ocp.cloud.ci.centos.org/blue/rest/organizations/jenkins/pipelines/mini-e2e_k8s-1.32/runs/299/nodes/90/log/?start=0)
```
Internal desc = rados: ret=-22, Invalid argument: "invalid value specified for ceph.dir.subvolume"


  I0730 13:40:52.897529       1 utils.go:341] ID: 52 Req-ID: pvc-8b39a787-726a-4bfb-88ae-51588a1d3376 GRPC call: /csi.v1.Controller/CreateVolume
  I0730 13:40:52.897608       1 utils.go:342] ID: 52 Req-ID: pvc-8b39a787-726a-4bfb-88ae-51588a1d3376 GRPC request: {"capacity_range":{"required_bytes":1073741824},"name":"pvc-8b39a787-726a-4bfb-88ae-51588a1d3376","parameters":{"clusterID":"9d5f2c8a-ea2f-4514-88a5-7f6b96f6e42a","csi.storage.k8s.io/pv/name":"pvc-8b39a787-726a-4bfb-88ae-51588a1d3376","csi.storage.k8s.io/pvc/name":"csi-cephfs-pvc","csi.storage.k8s.io/pvc/namespace":"cephfs-1210","fsName":"myfs","pool":"myfs-replicated"},"secrets":"***stripped***","volume_capabilities":[{"access_mode":{"mode":"MULTI_NODE_MULTI_WRITER"},"mount":{}}]}
  W0730 13:40:52.897630       1 credentials.go:57] adminID and adminKey are deprecated, please use userID and userKey instead
  I0730 13:40:52.900097       1 omap.go:89] ID: 52 Req-ID: pvc-8b39a787-726a-4bfb-88ae-51588a1d3376 got omap values: (pool="myfs-metadata", namespace="csi", name="csi.volumes.default"): map[]
  W0730 13:40:52.900119       1 credentials.go:57] adminID and adminKey are deprecated, please use userID and userKey instead
  I0730 13:40:52.909132       1 omap.go:159] ID: 52 Req-ID: pvc-8b39a787-726a-4bfb-88ae-51588a1d3376 set omap keys (pool="myfs-metadata", namespace="csi", name="csi.volumes.default"): map[csi.volume.pvc-8b39a787-726a-4bfb-88ae-51588a1d3376:a08ae0b0-6565-4fb7-8df7-23b9711f4fa7])
  I0730 13:40:52.912133       1 omap.go:159] ID: 52 Req-ID: pvc-8b39a787-726a-4bfb-88ae-51588a1d3376 set omap keys (pool="myfs-metadata", namespace="csi", name="csi.volume.a08ae0b0-6565-4fb7-8df7-23b9711f4fa7"): map[csi.imagename:csi-vol-a08ae0b0-6565-4fb7-8df7-23b9711f4fa7 csi.volname:pvc-8b39a787-726a-4bfb-88ae-51588a1d3376 csi.volume.owner:cephfs-1210])
  I0730 13:40:52.912157       1 fsjournal.go:320] ID: 52 Req-ID: pvc-8b39a787-726a-4bfb-88ae-51588a1d3376 Generated Volume ID (0001-0024-9d5f2c8a-ea2f-4514-88a5-7f6b96f6e42a-0000000000000001-a08ae0b0-6565-4fb7-8df7-23b9711f4fa7) and subvolume name (csi-vol-a08ae0b0-6565-4fb7-8df7-23b9711f4fa7) for request name (pvc-8b39a787-726a-4bfb-88ae-51588a1d3376)
  E0730 13:40:52.931939       1 volume.go:247] ID: 52 Req-ID: pvc-8b39a787-726a-4bfb-88ae-51588a1d3376 failed to create subvolume csi-vol-a08ae0b0-6565-4fb7-8df7-23b9711f4fa7 in fs myfs: rados: ret=-22, Invalid argument: "invalid value specified for ceph.dir.subvolume"
  E0730 13:40:52.931967       1 controllerserver.go:103] ID: 52 Req-ID: pvc-8b39a787-726a-4bfb-88ae-51588a1d3376 failed to create volume pvc-8b39a787-726a-4bfb-88ae-51588a1d3376: rados: ret=-22, Invalid argument: "invalid value specified for ceph.dir.subvolume"
  W0730 13:40:52.931983       1 credentials.go:57] adminID and adminKey are deprecated, please use userID and userKey instead
  I0730 13:40:52.937415       1 omap.go:126] ID: 52 Req-ID: pvc-8b39a787-726a-4bfb-88ae-51588a1d3376 removed omap keys (pool="myfs-metadata", namespace="csi", name="csi.volumes.default"): [csi.volume.pvc-8b39a787-726a-4bfb-88ae-51588a1d3376]
  E0730 13:40:52.937468       1 utils.go:346] ID: 52 Req-ID: pvc-8b39a787-726a-4bfb-88ae-51588a1d3376 GRPC error: rpc error: code = Internal desc = rados: ret=-22, Invalid argument: "invalid value specified for ceph.dir.subvolume"
```

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.32

Madhu-1 commented:
@iPraveenParihar please pick the new commit that need to cancel the clone and remove it for Already Exists error. Even i might take it as success and return non-error (not decided yet)

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.32

iPraveenParihar commented:
Local test results 

```
$ k get pvc
NAME                 STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   VOLUMEATTRIBUTESCLASS   AGE
cephfs-pvc           Bound    pvc-86611e26-4583-4702-9099-250832b0561d   1Gi        RWO            rook-cephfs    <unset>                 3m53s
cephfs-pvc-restore   Bound    pvc-fad915f4-1389-44f3-b8d8-e3e9226a0b83   1Gi        ROX            rook-cephfs    <unset>                 3m6s
```

- NodeStageVolume
```
I0811 08:53:44.241188       1 utils.go:341] ID: 48 Req-ID: 0001-0009-rook-ceph-0000000000000001-a446c7b7-238a-472e-b75e-4c9b729ba230 GRPC call: /csi.v1.Node/NodeStageVolume
I0811 08:53:44.242061       1 utils.go:342] ID: 48 Req-ID: 0001-0009-rook-ceph-0000000000000001-a446c7b7-238a-472e-b75e-4c9b729ba230 GRPC request: {"secrets":"***stripped***","staging_target_path":"/var/lib/kubelet/plugins/kubernetes.io/csi/rook-ceph.cephfs.csi.ceph.com/4dcaf4179e430c045094d39c44073dab54bf210f4b8eeabca8b8bbe89837f6eb/globalmount","volume_capability":{"access_mode":{"mode":"SINGLE_NODE_MULTI_WRITER"},"mount":{}},"volume_context":{"clusterID":"rook-ceph","fsName":"myfs","pool":"myfs-replicated","storage.kubernetes.io/csiProvisionerIdentity":"1754901136768-6917-rook-ceph.cephfs.csi.ceph.com","subvolumeName":"csi-vol-a446c7b7-238a-472e-b75e-4c9b729ba230","subvolumePath":"/volumes/csi/csi-vol-a446c7b7-238a-472e-b75e-4c9b729ba230/0510e7e2-27ba-4579-8822-a45d57b65f30"},"volume_id":"0001-0009-rook-ceph-0000000000000001-a446c7b7-238a-472e-b75e-4c9b729ba230"}
I0811 08:53:44.254950       1 omap.go:89] ID: 48 Req-ID: 0001-0009-rook-ceph-0000000000000001-a446c7b7-238a-472e-b75e-4c9b729ba230 got omap values: (pool="myfs-metadata", namespace="csi", name="csi.volume.a446c7b7-238a-472e-b75e-4c9b729ba230"): map[csi.imagename:csi-vol-a446c7b7-238a-472e-b75e-4c9b729ba230 csi.volname:pvc-adaa9844-309f-450f-b3b5-1a46ec82ba5b csi.volume.owner:rook-ceph]
I0811 08:53:44.314553       1 nodeserver.go:389] ID: 48 Req-ID: 0001-0009-rook-ceph-0000000000000001-a446c7b7-238a-472e-b75e-4c9b729ba230 user ID mapping metadata .cephfs.csi.ceph.com/userID/0001-0009-rook-ceph-0000000000000001-a446c7b7-238a-472e-b75e-4c9b729ba230/c1 set for subvolume csi-vol-a446c7b7-238a-472e-b75e-4c9b729ba230
I0811 08:53:44.315086       1 nodeserver.go:464] ID: 48 Req-ID: 0001-0009-rook-ceph-0000000000000001-a446c7b7-238a-472e-b75e-4c9b729ba230 cephfs: mounting volume 0001-0009-rook-ceph-0000000000000001-a446c7b7-238a-472e-b75e-4c9b729ba230 with Ceph kernel client
I0811 08:53:44.456182       1 cephcmds.go:109] ID: 48 Req-ID: 0001-0009-rook-ceph-0000000000000001-a446c7b7-238a-472e-b75e-4c9b729ba230 command succeeded: mount [-t ceph csi-cephfs-node@735a80cd-a89c-4a1a-bed8-5cf157707df0.myfs=/volumes/csi/csi-vol-a446c7b7-238a-472e-b75e-4c9b729ba230/0510e7e2-27ba-4579-8822-a45d57b65f30 /var/lib/kubelet/plugins/kubernetes.io/csi/rook-ceph.cephfs.csi.ceph.com/4dcaf4179e430c045094d39c44073dab54bf210f4b8eeabca8b8bbe89837f6eb/globalmount -o mon_addr=10.108.245.6:6789,secretfile=/tmp/csi/keys/keyfile-146122425,_netdev]
I0811 08:53:44.456376       1 nodeserver.go:314] ID: 48 Req-ID: 0001-0009-rook-ceph-0000000000000001-a446c7b7-238a-472e-b75e-4c9b729ba230 cephfs: successfully mounted volume 0001-0009-rook-ceph-0000000000000001-a446c7b7-238a-472e-b75e-4c9b729ba230 to /var/lib/kubelet/plugins/kubernetes.io/csi/rook-ceph.cephfs.csi.ceph.com/4dcaf4179e430c045094d39c44073dab54bf210f4b8eeabca8b8bbe89837f6eb/globalmount
I0811 08:53:44.456676       1 utils.go:348] ID: 48 Req-ID: 0001-0009-rook-ceph-0000000000000001-a446c7b7-238a-472e-b75e-4c9b729ba230 GRPC response: {}

///////////////////// Shallow Volume /////////////////////

I0811 08:54:35.779058       1 utils.go:341] ID: 58 Req-ID: 0001-0009-rook-ceph-0000000000000001-b29f42f5-4b3a-4337-8ade-5ae1c788834f GRPC call: /csi.v1.Node/NodeStageVolume
I0811 08:54:35.779470       1 utils.go:342] ID: 58 Req-ID: 0001-0009-rook-ceph-0000000000000001-b29f42f5-4b3a-4337-8ade-5ae1c788834f GRPC request: {"secrets":"***stripped***","staging_target_path":"/var/lib/kubelet/plugins/kubernetes.io/csi/rook-ceph.cephfs.csi.ceph.com/f2d96580311a3167a7c9a579cf7c697a0b5ac441b951de10f9a9bf6c2a0d280d/globalmount","volume_capability":{"access_mode":{"mode":"MULTI_NODE_READER_ONLY"},"mount":{}},"volume_context":{"clusterID":"rook-ceph","fsName":"myfs","pool":"myfs-replicated","storage.kubernetes.io/csiProvisionerIdentity":"1754901136768-6917-rook-ceph.cephfs.csi.ceph.com","subvolumeName":"csi-vol-b29f42f5-4b3a-4337-8ade-5ae1c788834f","subvolumePath":"/volumes/csi/csi-vol-a446c7b7-238a-472e-b75e-4c9b729ba230/"},"volume_id":"0001-0009-rook-ceph-0000000000000001-b29f42f5-4b3a-4337-8ade-5ae1c788834f"}
I0811 08:54:35.787627       1 omap.go:89] ID: 58 Req-ID: 0001-0009-rook-ceph-0000000000000001-b29f42f5-4b3a-4337-8ade-5ae1c788834f got omap values: (pool="myfs-metadata", namespace="csi", name="csi.volume.b29f42f5-4b3a-4337-8ade-5ae1c788834f"): map[csi.imagename:csi-vol-b29f42f5-4b3a-4337-8ade-5ae1c788834f csi.volname:pvc-c7002a64-92fa-4a88-8e20-5ee1cf5b5cab csi.volume.backingsnapshotid:0001-0009-rook-ceph-0000000000000001-6f3f4137-6a8a-4de2-9749-77992a19ca0b csi.volume.owner:rook-ceph]
I0811 08:54:35.796395       1 omap.go:89] ID: 58 Req-ID: 0001-0009-rook-ceph-0000000000000001-b29f42f5-4b3a-4337-8ade-5ae1c788834f got omap values: (pool="myfs-metadata", namespace="csi", name="csi.snap.6f3f4137-6a8a-4de2-9749-77992a19ca0b"): map[csi.imagename:csi-snap-6f3f4137-6a8a-4de2-9749-77992a19ca0b csi.snapname:snapshot-c90ebf87-8cb2-471f-bc19-5ad98ab9abbd csi.source:csi-vol-a446c7b7-238a-472e-b75e-4c9b729ba230]
I0811 08:54:35.842228       1 omap.go:89] ID: 58 Req-ID: 0001-0009-rook-ceph-0000000000000001-b29f42f5-4b3a-4337-8ade-5ae1c788834f got omap values: (pool="myfs-metadata", namespace="csi", name="csi.snap.6f3f4137-6a8a-4de2-9749-77992a19ca0b"): map[csi.imagename:csi-snap-6f3f4137-6a8a-4de2-9749-77992a19ca0b csi.snapname:snapshot-c90ebf87-8cb2-471f-bc19-5ad98ab9abbd csi.source:csi-vol-a446c7b7-238a-472e-b75e-4c9b729ba230]
I0811 08:54:35.909810       1 nodeserver.go:377] ID: 58 Req-ID: 0001-0009-rook-ceph-0000000000000001-b29f42f5-4b3a-4337-8ade-5ae1c788834f user ID mapping metadata .cephfs.csi.ceph.com/userID/0001-0009-rook-ceph-0000000000000001-b29f42f5-4b3a-4337-8ade-5ae1c788834f/c1 set for subvolume snapshot csi-snap-6f3f4137-6a8a-4de2-9749-77992a19ca0b
I0811 08:54:35.910404       1 nodeserver.go:464] ID: 58 Req-ID: 0001-0009-rook-ceph-0000000000000001-b29f42f5-4b3a-4337-8ade-5ae1c788834f cephfs: mounting volume 0001-0009-rook-ceph-0000000000000001-b29f42f5-4b3a-4337-8ade-5ae1c788834f with Ceph kernel client
I0811 08:54:36.010954       1 cephcmds.go:109] ID: 58 Req-ID: 0001-0009-rook-ceph-0000000000000001-b29f42f5-4b3a-4337-8ade-5ae1c788834f command succeeded: mount [-t ceph csi-cephfs-node@735a80cd-a89c-4a1a-bed8-5cf157707df0.myfs=/volumes/csi/csi-vol-a446c7b7-238a-472e-b75e-4c9b729ba230/ /var/lib/kubelet/plugins/kubernetes.io/csi/rook-ceph.cephfs.csi.ceph.com/f2d96580311a3167a7c9a579cf7c697a0b5ac441b951de10f9a9bf6c2a0d280d/globalmount -o mon_addr=10.108.245.6:6789,secretfile=/tmp/csi/keys/keyfile-3760381552,ro,_netdev]
I0811 08:54:36.025641       1 cephcmds.go:109] ID: 58 Req-ID: 0001-0009-rook-ceph-0000000000000001-b29f42f5-4b3a-4337-8ade-5ae1c788834f command succeeded: mount [-o bind,_netdev /var/lib/kubelet/plugins/kubernetes.io/csi/rook-ceph.cephfs.csi.ceph.com/f2d96580311a3167a7c9a579cf7c697a0b5ac441b951de10f9a9bf6c2a0d280d/globalmount/.snap/csi-snap-6f3f4137-6a8a-4de2-9749-77992a19ca0b/0510e7e2-27ba-4579-8822-a45d57b65f30 /var/lib/kubelet/plugins/kubernetes.io/csi/rook-ceph.cephfs.csi.ceph.com/f2d96580311a3167a7c9a579cf7c697a0b5ac441b951de10f9a9bf6c2a0d280d/globalmount]
I0811 08:54:36.033355       1 cephcmds.go:109] ID: 58 Req-ID: 0001-0009-rook-ceph-0000000000000001-b29f42f5-4b3a-4337-8ade-5ae1c788834f command succeeded: mount [-o bind,_netdev,remount /var/lib/kubelet/plugins/kubernetes.io/csi/rook-ceph.cephfs.csi.ceph.com/f2d96580311a3167a7c9a579cf7c697a0b5ac441b951de10f9a9bf6c2a0d280d/globalmount]
I0811 08:54:36.033584       1 nodeserver.go:314] ID: 58 Req-ID: 0001-0009-rook-ceph-0000000000000001-b29f42f5-4b3a-4337-8ade-5ae1c788834f cephfs: successfully mounted volume 0001-0009-rook-ceph-0000000000000001-b29f42f5-4b3a-4337-8ade-5ae1c788834f to /var/lib/kubelet/plugins/kubernetes.io/csi/rook-ceph.cephfs.csi.ceph.com/f2d96580311a3167a7c9a579cf7c697a0b5ac441b951de10f9a9bf6c2a0d280d/globalmount
I0811 08:54:36.033721       1 utils.go:348] ID: 58 Req-ID: 0001-0009-rook-ceph-0000000000000001-b29f42f5-4b3a-4337-8ade-5ae1c788834f GRPC response: {}
```

- ControllerUnpublishVolume
```
I0811 08:57:18.523177       1 utils.go:341] ID: 38 GRPC call: /csi.v1.Controller/ControllerUnpublishVolume
I0811 08:57:18.524698       1 utils.go:342] ID: 38 GRPC request: {"node_id":"c1","secrets":"***stripped***","volume_id":"0001-0009-rook-ceph-0000000000000001-a446c7b7-238a-472e-b75e-4c9b729ba230"}
I0811 08:57:18.539971       1 omap.go:89] ID: 38 got omap values: (pool="myfs-metadata", namespace="csi", name="csi.volume.a446c7b7-238a-472e-b75e-4c9b729ba230"): map[csi.imagename:csi-vol-a446c7b7-238a-472e-b75e-4c9b729ba230 csi.volname:pvc-adaa9844-309f-450f-b3b5-1a46ec82ba5b csi.volume.owner:rook-ceph]
I0811 08:57:18.676241       1 controllerserver.go:1287] ID: 38 user ID mapping metadata .cephfs.csi.ceph.com/userID/0001-0009-rook-ceph-0000000000000001-a446c7b7-238a-472e-b75e-4c9b729ba230/c1 unset for subvolume csi-vol-a446c7b7-238a-472e-b75e-4c9b729ba230
I0811 08:57:18.676485       1 utils.go:348] ID: 38 GRPC response: {}

///////////////////// Shallow Volume /////////////////////

I0811 08:57:18.534649       1 utils.go:341] ID: 39 GRPC call: /csi.v1.Controller/ControllerUnpublishVolume
I0811 08:57:18.535147       1 utils.go:342] ID: 39 GRPC request: {"node_id":"c1","secrets":"***stripped***","volume_id":"0001-0009-rook-ceph-0000000000000001-b29f42f5-4b3a-4337-8ade-5ae1c788834f"}
I0811 08:57:18.544983       1 omap.go:89] ID: 39 got omap values: (pool="myfs-metadata", namespace="csi", name="csi.volume.b29f42f5-4b3a-4337-8ade-5ae1c788834f"): map[csi.imagename:csi-vol-b29f42f5-4b3a-4337-8ade-5ae1c788834f csi.volname:pvc-c7002a64-92fa-4a88-8e20-5ee1cf5b5cab csi.volume.backingsnapshotid:0001-0009-rook-ceph-0000000000000001-6f3f4137-6a8a-4de2-9749-77992a19ca0b csi.volume.owner:rook-ceph]
I0811 08:57:18.553608       1 omap.go:89] ID: 39 got omap values: (pool="myfs-metadata", namespace="csi", name="csi.snap.6f3f4137-6a8a-4de2-9749-77992a19ca0b"): map[csi.imagename:csi-snap-6f3f4137-6a8a-4de2-9749-77992a19ca0b csi.snapname:snapshot-c90ebf87-8cb2-471f-bc19-5ad98ab9abbd csi.source:csi-vol-a446c7b7-238a-472e-b75e-4c9b729ba230]
I0811 08:57:18.669737       1 omap.go:89] ID: 39 got omap values: (pool="myfs-metadata", namespace="csi", name="csi.snap.6f3f4137-6a8a-4de2-9749-77992a19ca0b"): map[csi.imagename:csi-snap-6f3f4137-6a8a-4de2-9749-77992a19ca0b csi.snapname:snapshot-c90ebf87-8cb2-471f-bc19-5ad98ab9abbd csi.source:csi-vol-a446c7b7-238a-472e-b75e-4c9b729ba230]
I0811 08:57:18.805848       1 controllerserver.go:1275] ID: 39 user ID mapping metadata .cephfs.csi.ceph.com/userID/0001-0009-rook-ceph-0000000000000001-b29f42f5-4b3a-4337-8ade-5ae1c788834f/c1 unset for subvolume snapshot csi-snap-6f3f4137-6a8a-4de2-9749-77992a19ca0b
I0811 08:57:18.806441       1 utils.go:348] ID: 39 GRPC response: {}
```

```
$ ceph fs subvolume metadata ls myfs csi-vol-a446c7b7-238a-472e-b75e-4c9b729ba230 csi
{
    ".cephfs.csi.ceph.com/userid/0001-0009-rook-ceph-0000000000000001-a446c7b7-238a-472e-b75e-4c9b729ba230/c1": "csi-cephfs-node",
    "csi.storage.k8s.io/pv/name": "pvc-adaa9844-309f-450f-b3b5-1a46ec82ba5b",
    "csi.storage.k8s.io/pvc/name": "cephfs-pvc",
    "csi.storage.k8s.io/pvc/namespace": "rook-ceph"
}

$ ceph fs subvolume snapshot metadata ls myfs csi-vol-a446c7b7-238a-472e-b75e-4c9b729ba230 csi-snap-6f3f4137-6a8a-4de2-9749-77992a19ca0b csi
{
    ".cephfs.csi.ceph.com/userid/0001-0009-rook-ceph-0000000000000001-b29f42f5-4b3a-4337-8ade-5ae1c788834f/c1": "csi-cephfs-node",
    "csi.storage.k8s.io/volumesnapshot/name": "cephfs-pvc-snapshot",
    "csi.storage.k8s.io/volumesnapshot/namespace": "rook-ceph",
    "csi.storage.k8s.io/volumesnapshotcontent/name": "snapcontent-c90ebf87-8cb2-471f-bc19-5ad98ab9abbd"
}
```

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.32

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.32

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.33

iPraveenParihar commented:
- ceph.dir.volume error [logs](https://jenkins-ceph-csi.apps.ocp.cloud.ci.centos.org/blue/rest/organizations/jenkins/pipelines/mini-e2e_k8s-1.32/runs/328/nodes/90/log/?start=0)

```
  I0811 11:07:20.839331       1 utils.go:341] ID: 53 Req-ID: pvc-56e42521-767b-4db6-bbd3-a167c869c4d8 GRPC call: /csi.v1.Controller/CreateVolume
  I0811 11:07:20.839426       1 utils.go:342] ID: 53 Req-ID: pvc-56e42521-767b-4db6-bbd3-a167c869c4d8 GRPC request: {"capacity_range":{"required_bytes":1073741824},"name":"pvc-56e42521-767b-4db6-bbd3-a167c869c4d8","parameters":{"clusterID":"15d8c720-35d0-4986-8905-745e9a386ce4","csi.storage.k8s.io/pv/name":"pvc-56e42521-767b-4db6-bbd3-a167c869c4d8","csi.storage.k8s.io/pvc/name":"csi-cephfs-pvc","csi.storage.k8s.io/pvc/namespace":"cephfs-7673","fsName":"myfs","pool":"myfs-replicated"},"secrets":"***stripped***","volume_capabilities":[{"access_mode":{"mode":"MULTI_NODE_MULTI_WRITER"},"mount":{}}]}
  W0811 11:07:20.839498       1 credentials.go:57] adminID and adminKey are deprecated, please use userID and userKey instead
  I0811 11:07:20.842305       1 omap.go:89] ID: 53 Req-ID: pvc-56e42521-767b-4db6-bbd3-a167c869c4d8 got omap values: (pool="myfs-metadata", namespace="csi", name="csi.volumes.default"): map[]
  W0811 11:07:20.842341       1 credentials.go:57] adminID and adminKey are deprecated, please use userID and userKey instead
  I0811 11:07:20.847704       1 omap.go:159] ID: 53 Req-ID: pvc-56e42521-767b-4db6-bbd3-a167c869c4d8 set omap keys (pool="myfs-metadata", namespace="csi", name="csi.volumes.default"): map[csi.volume.pvc-56e42521-767b-4db6-bbd3-a167c869c4d8:e8596899-a6ab-4c71-99b9-4520db44f4ca])
  I0811 11:07:20.850552       1 omap.go:159] ID: 53 Req-ID: pvc-56e42521-767b-4db6-bbd3-a167c869c4d8 set omap keys (pool="myfs-metadata", namespace="csi", name="csi.volume.e8596899-a6ab-4c71-99b9-4520db44f4ca"): map[csi.imagename:csi-vol-e8596899-a6ab-4c71-99b9-4520db44f4ca csi.volname:pvc-56e42521-767b-4db6-bbd3-a167c869c4d8 csi.volume.owner:cephfs-7673])
  I0811 11:07:20.850584       1 fsjournal.go:320] ID: 53 Req-ID: pvc-56e42521-767b-4db6-bbd3-a167c869c4d8 Generated Volume ID (0001-0024-15d8c720-35d0-4986-8905-745e9a386ce4-0000000000000001-e8596899-a6ab-4c71-99b9-4520db44f4ca) and subvolume name (csi-vol-e8596899-a6ab-4c71-99b9-4520db44f4ca) for request name (pvc-56e42521-767b-4db6-bbd3-a167c869c4d8)
  E0811 11:07:20.865552       1 volume.go:247] ID: 53 Req-ID: pvc-56e42521-767b-4db6-bbd3-a167c869c4d8 failed to create subvolume csi-vol-e8596899-a6ab-4c71-99b9-4520db44f4ca in fs myfs: rados: ret=-22, Invalid argument: "invalid value specified for ceph.dir.subvolume"
  E0811 11:07:20.865586       1 controllerserver.go:103] ID: 53 Req-ID: pvc-56e42521-767b-4db6-bbd3-a167c869c4d8 failed to create volume pvc-56e42521-767b-4db6-bbd3-a167c869c4d8: rados: ret=-22, Invalid argument: "invalid value specified for ceph.dir.subvolume"
  W0811 11:07:20.865617       1 credentials.go:57] adminID and adminKey are deprecated, please use userID and userKey instead
  I0811 11:07:20.870728       1 omap.go:126] ID: 53 Req-ID: pvc-56e42521-767b-4db6-bbd3-a167c869c4d8 removed omap keys (pool="myfs-metadata", namespace="csi", name="csi.volumes.default"): [csi.volume.pvc-56e42521-767b-4db6-bbd3-a167c869c4d8]
  E0811 11:07:20.870794       1 utils.go:346] ID: 53 Req-ID: pvc-56e42521-767b-4db6-bbd3-a167c869c4d8 GRPC error: rpc error: code = Internal desc = rados: ret=-22, Invalid argument: "invalid value specified for ceph.dir.subvolume"
 ```

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.33

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.33

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.33

Madhu-1 commented on e2e/cephfs.go:
```diff
@@ -456,19 +476,21 @@ var _ = Describe(cephfsType, func() {
 				validateSubvolumeCount(f, 1, fileSystemName, subvolumegroup)
 				validateOmapCount(f, 1, cephfsType, metadataPool, volumesType)
-				pvcName := app.Spec.Volumes[0].Name
+				pvcName := fmt.Sprintf("%s-%s", app.Name, app.Spec.Volumes[0].Name)
```
why there is a change in existing testcase? if requried lets create a new test case for it or update the `verifyUserIdMappingMetadata` to take care of all the different combination as its same for both cephfs and RBD

iPraveenParihar replied:
This changes is required because waitForPVCToBeDeleted changed to waitForPVToBeDeleted which requires the PV name to be passed.

iPraveenParihar replied:
Also Previously, it was incorrect.

This is an `ephemeral volume support` test case, where the PVC name should be `<pod-name>-<pod.spec.volumes[*].name>`. However, we were mistakenly waiting for the PVC named only `pod.spec.volumes[*].name`, which doesn’t exist.

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.33

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.33

iPraveenParihar commented:
Had some changes split into different commit, taken care now.
@Rakshith-R PTAL

Madhu-1 commented on e2e/pod.go:
```diff
@@ -738,3 +747,155 @@ func verifyReadAffinity(
+	nodeId := pod.Spec.NodeName
+	_, pvObject, err := getPVCAndPV(f.ClientSet, pvc.Name, pvc.Namespace)
+	if err != nil {
+		framework.Logf("error getting pvc %q in namespace %q: %v", pvc.Name, pvc.Namespace, err)
```
this is not required if more details are required add it to error message

Madhu-1 commented on e2e/pvc.go:
```diff
@@ -305,7 +305,7 @@ func deletePVCAndValidatePV(c kubernetes.Interface, pvc *v1.PersistentVolumeClai
 
 	err = c.CoreV1().PersistentVolumeClaims(nameSpace).Delete(ctx, name, metav1.DeleteOptions{})
-	if err != nil {
+	if err != nil && !apierrs.IsNotFound(err) {
```
when this can happen?

iPraveenParihar replied:
added by mistake, not needed removed it.

Madhu-1 commented on e2e/pvc.go:
```diff
@@ -453,31 +453,31 @@ func getMetricsForPVC(f *framework.Framework, pvc *v1.PersistentVolumeClaim, t i
 }
 
-func waitForPVCToBeDeleted(c kubernetes.Interface, namespace, pvcName string, t int) error {
+func waitForPVToBeDeleted(c kubernetes.Interface, pvName string, t int) error {
```
Not able to understand the reason for this change

iPraveenParihar replied:
For 'ephemeral volume support' test case (may be for other too), PVC is deleted but PV is deleted after some delay. (https://github.com/ceph/ceph-csi/pull/5445/files#r2235638823)
So, instead of waiting for PVC to be deleted changed it to wait for PV to be deleted.

Madhu-1 replied:
is this happening because of this PR? we never had any problem that required this change.

nixpanic replied:
Sounds like this is unrelated to the metadata storing. Could this be a cause for left-over volumes after some tests? Might be better suitable in it's own PR.

Madhu-1 replied:
Please revert back this change

iPraveenParihar replied:
@Madhu-1, with ControllerPublish support csi-attacher will add an finalizer to the PV and will wait until ControllerUnpublishVolume completes before removing the finalizer. This can leave PV object to be present after the PVC is deleted. 

So the change was to wait for PV object instead of PVC object.

Madhu-1 replied:
deletePVCAndValidatePV does the same thing, it waits for PV deletion not just for PVC deletion.

Madhu-1 commented on e2e/utils.go:
```diff
@@ -582,6 +582,11 @@ func validatePVCAndAppWaitForFirstConsumer(pvcPath, appPath string, f *framework
 		return err
 	}
 
+	_, err = getBoundPV(f.ClientSet, pvc)
```
why do we need this check?

iPraveenParihar replied:
not needed removed it.

Madhu-1 commented on internal/cephfs/controllerserver.go:
```diff
@@ -1143,12 +1143,32 @@ func (cs *ControllerServer) ControllerPublishVolume(
+	if err != nil {
+		log.WarningLog(ctx, "controller publish secret not found: %v", err)
+
+		return &csi.ControllerPublishVolumeResponse{}, nil
```
i think this can be a wrong assumption because we might have failed to read the secret from the api-server, we should return error only when secret is not found

iPraveenParihar replied:
Okay, I'll have to move the fetching k8s secrets code here.

Madhu-1 replied:
Please return known error message and do compare based on the return error or success

iPraveenParihar replied:
The K8s API can return many different kinds of errors, so instead of comparing raw error messages I’ve moved the k8s secret fetching logic outside for now.

Madhu-1 replied:
we need specific error check for `apierror.NotFound(err) || errors.Is(err.NosecretsFound)` nothing else, if there are any other errors return error from here or else return success, with new changes i see lot of duplicate code which can be avoided

iPraveenParihar replied:
In `GetControllerPublishSecretRef`, we also do `DecomposeCSIID` which will fail for static volume. (we don't have clusterId information from ControllerUnpublishVolumeRequest).

So, moving k8s API call out is better?

Madhu-1 replied:
Please add a TODO to clean it up later.

Madhu-1 commented on internal/cephfs/controllerserver.go:
```diff
@@ -1173,26 +1193,110 @@ func (cs *ControllerServer) ControllerUnpublishVolume(
+			volOptions.SubVolume.VolID, err)
+	}
+
+	log.DebugLog(ctx, "userID mapping metadata %s unset for subvolume %s",
```
we are logging the volumeID not the subvolume?

iPraveenParihar replied:
`VolID` is the subvolume name actually

```
I0813 11:29:36.229232       1 utils.go:341] ID: 21 GRPC call: /csi.v1.Controller/ControllerUnpublishVolume
I0813 11:29:36.229577       1 utils.go:342] ID: 21 GRPC request: {"node_id":"c1","secrets":"***stripped***","volume_id":"0001-0009-rook-ceph-0000000000000001-26224448-2aba-4aba-aa43-ad6e29b6f99e"}
W0813 11:29:36.229963       1 credentials.go:57] adminID and adminKey are deprecated, please use userID and userKey instead
W0813 11:29:36.231475       1 credentials.go:57] adminID and adminKey are deprecated, please use userID and userKey instead
I0813 11:29:36.243413       1 omap.go:89] ID: 21 got omap values: (pool="myfs-metadata", namespace="csi", name="csi.volume.26224448-2aba-4aba-aa43-ad6e29b6f99e"): map[csi.imagename:csi-vol-26224448-2aba-4aba-aa43-ad6e29b6f99e csi.volname:pvc-8a53dfc3-f090-44be-aa13-4f89563caa3f csi.volume.owner:rook-ceph]
I0813 11:29:36.432838       1 controllerserver.go:1304] ID: 21 userID mapping metadata .cephfs.csi.ceph.com/userid/0001-0009-rook-ceph-0000000000000001-26224448-2aba-4aba-aa43-ad6e29b6f99e/c1 unset for subvolume csi-vol-26224448-2aba-4aba-aa43-ad6e29b6f99e
I0813 11:29:36.433409       1 utils.go:348] ID: 21 GRPC response: {}
```

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -1705,12 +1705,33 @@ func (cs *ControllerServer) ControllerPublishVolume(
+	if err != nil {
+		log.WarningLog(ctx, "controller publish secret not found: %v", err)
+
+		return &csi.ControllerPublishVolumeResponse{}, nil
```
same problem here with error handling, This can lead to problems

Madhu-1 commented on internal/rbd/nodeserver.go:
```diff
@@ -55,6 +55,9 @@ type NodeServer struct {
 	xfsHasReflinkSupport featureFlag
+
+	// Set metadata on volume
+	SetMetadata bool
```
this can be small case?

iPraveenParihar replied:
I can do it only for CephFs as the `SetMetadata`  is driver.go.
For CephFs this would be in same package (internal/cephfs)
But for RBD this is in different package (internal/rbd/driver)

Madhu-1 reviewed (commented):
i see one bug with error handling, other things LGTM

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.33

Madhu-1 commented on e2e/rbd.go:
```diff
@@ -907,18 +914,20 @@ var _ = Describe("RBD", func() {
 				validateRBDImageCount(f, 1, defaultRBDPool)
 				validateOmapCount(f, 1, rbdType, defaultRBDPool, volumesType)
-				pvcName := app.Spec.Volumes[0].Name
+				pvcName := fmt.Sprintf("%s-%s", app.Name, app.Spec.Volumes[0].Name)
```
please revert back this change as this is nothing to do with this PR

iPraveenParihar replied:
This changes is required because `waitForPVCToBeDeleted` changed to `waitForPVToBeDeleted` which requires the PV name to be passed.

nixpanic commented on internal/cephfs/driver.go:
```diff
@@ -141,8 +142,8 @@ func (fs *Driver) Run(conf *util.Config) {
-		// if fencing is enabled, we can add the PUBLISH_UNPUBLISH_VOLUME capability.
-		if fs.cd.IsFencingEnabled() {
+		// if fencing or setmetadata flag is enabled, we need PUBLISH_UNPUBLISH_VOLUME capability.
+		if fs.cd.IsFencingEnabled() || conf.SetMetadata {
```
Did you not mention that `ControllerServiceCapability_RPC_PUBLISH_UNPUBLISH_VOLUME` should always be enabled? After disabling, it was possible that VolumeAttachments are not cleaned up?

iPraveenParihar replied:
Yes, that is taken care in different PR - https://github.com/ceph/ceph-csi/pull/5503
Will rebase this once the bug fix PR is fixed.

iPraveenParihar committed:
rbd: set/unset userIdMapping matadata

This commit does the following operations:

1. In NodeStageVolume(): sets nodeId:userId mapping in image metadata.
   The userId is the ceph user used for mounting the subvolume.

2. In ControllerUnpulishVolume(): removes the nodeId:userId mapping from
   the image metadata.

Signed-off-by: Praveen M <m.praveen@ibm.com>

iPraveenParihar committed:
cephfs: set/unset userIdMapping metadata

This commit does the following operations:

1. In NodeStageVolume(): sets nodeId:userId mapping in subvolume metadata.
   The userId is the ceph user used for mounting the subvolume.

2. In ControllerUnpulishVolume(): removes the nodeId:userId mapping from
   the subvolume metadata.

Signed-off-by: Praveen M <m.praveen@ibm.com>

iPraveenParihar committed:
e2e: add tests to verify userid mapping metadata

Signed-off-by: Praveen M <m.praveen@ibm.com>

iPraveenParihar committed:
doc: update PendingReleaseNotes.md

Signed-off-by: Praveen M <m.praveen@ibm.com>

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.33

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.33

iPraveenParihar commented:
passed CI
- https://jenkins-ceph-csi.apps.ocp.cloud.ci.centos.org/blue/organizations/jenkins/mini-e2e-helm_k8s-1.33/detail/mini-e2e-helm_k8s-1.33/112/pipeline
- https://jenkins-ceph-csi.apps.ocp.cloud.ci.centos.org/blue/organizations/jenkins/mini-e2e_k8s-1.33/detail/mini-e2e_k8s-1.33/116/pipeline

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.33

Rakshith-R reviewed (approved):
Thanks

Rakshith-R commented:
@mergifyio queue
